### PR TITLE
libblkid: bitlocker: read uuid and label

### PR DESCRIPTION
**Pull request by viccie30**
These changes use the FVE metadata to read a GUID and a label from BitLocker partitions. 

It might be possible to share the parsing code for Microsoft GUIDs with libfdisk, which has similar code at https://github.com/util-linux/util-linux/blob/2d07a4deb71aa564d740b8f3c82c8808431a0ccf/libfdisk/src/gpt.c#L226-L255. A [comment in libcryptsetup](https://gitlab.com/cryptsetup/cryptsetup/-/blob/main/lib/bitlk/bitlk.c?ref_type=heads#L67-86) suggests adding this in libuuid, but I have not attempted that.

### CODE REVIEW

**Comment by t-8ch** [comment]
Do you have example images?
If so, could you provide a testcase as the first commit?

**Comment by viccie30** [comment]
I will see if I can make some images using a Windows image

**Comment by karelzak** [comment]
I think it's okay to use `blkid_probe_sprintf_uuid()` rather than reuse the code from libfdisk, as this library uses libuuid to parse the uuids. We depended on libuuid in libblkid, but it was removed years ago to reduce the dependencies.

File: libblkid/src/superblocks/bitlocker.c [hunk 1]
```diff
@@ -149,26 +182,50 @@ static int probe_bitlocker(blkid_probe pr,
...
+			"%02x%02x%02x%02x-%02x%02x-%02x%02x-%02x%02x-%02x%02x%02x%02x%02x%02x",
+			fve->header.volume_identifier[3], fve->header.volume_identifier[2], /* uint32_t Data1 */
+			fve->header.volume_identifier[1], fve->header.volume_identifier[0],
+			fve->header.volume_identifier[5], fve->header.volume_identifier[4], /* uint16_t Data2 */
+			fve->header.volume_identifier[7], fve->header.volume_identifier[6], /* uint16_t Data3 */
+			fve->header.volume_identifier[8], fve->header.volume_identifier[9], /* uint8_t Data4[8] */
+			fve->header.volume_identifier[10], fve->header.volume_identifier[11],
+			fve->header.volume_identifier[12], fve->header.volume_identifier[13],
+			fve->header.volume_identifier[14], fve->header.volume_identifier[15]);
+
```
[the remark this answers is not in the document]
Comment (viccie30) [reply libblkid/src/superblocks/bitlocker.c]: I've made this change.

**Comment by viccie30** [comment]
Looking at the [logs](https://download.copr.fedorainfracloud.org/results/packit/util-linux-util-linux-3135/fedora-rawhide-i386/07788679-util-linux/builder-live.log.gz) the build failure on i386 looks like a failure in the builder and unrelated to the changes.

**Comment by t-8ch** [comment]
> failure in the builder

Indeed. But there are many more CI checks, which @karelzak needs to approve to run first.

File: libblkid/src/superblocks/bitlocker.c [hunk 2]
```diff
@@ -38,10 +38,33 @@ struct bde_header_togo {
...
 
 
-struct bde_fve_metadata {
+struct bde_fve_metadata_block_header {
 /*   0 */ unsigned char  signature[8];
-/*   8 */ uint16_t       size;
+/*   8 */ unsigned char  __dummy1[10 - 8];
 /*  10 */ uint16_t       version;
+/*  12 */ unsigned char  __dummy2[64 - 12];
+};
```
Comment (t-8ch) [inline libblkid/src/superblocks/bitlocker.c:46]: These should all be packed.

File: libblkid/src/superblocks/bitlocker.c [hunk 3]
```diff
@@ -149,26 +182,50 @@ static int probe_bitlocker(blkid_probe pr,
...
+		for (off = sizeof(struct bde_fve_metadata_header);
+		     off + sizeof(struct bde_fve_metadata_entry) < le32_to_cpu(fve->header.size);
+		     off += le16_to_cpu(entry->size)) {
+			entry = (const struct bde_fve_metadata_entry *) ((const char *) &fve->header + off);
+			if (off % 2 ||
+			    le16_to_cpu(entry->size) < sizeof(struct bde_fve_metadata_entry) ||
+			    off + le16_to_cpu(entry->size) >= le32_to_cpu(fve->header.size))
+				return -1;
+
+			if (entry->entry_type == BDE_METADATA_ENTRY_TYPE_DESCRIPTION &&
```
Comment (t-8ch) [inline libblkid/src/superblocks/bitlocker.c:207]: Here the `cpu_to_le16()` is part of `BDE_METADATA_ENTRY_TYPE_DESCRIPTION`.
Doing the endianess conversion as part of  `le16_to_cpu(entry->entry_type)` seems more uniform with the rest of the code.

File: tests/expected/blkid/low-probe-bitlocker-win7
Comment (t-8ch) [inline tests/expected/blkid/low-probe-bitlocker-win7:1]: Thanks!

File: libblkid/src/superblocks/bitlocker.c [hunk 4]
```diff
@@ -149,26 +182,50 @@ static int probe_bitlocker(blkid_probe pr,
...
-				(const unsigned char *) &hdr->volume_serial,
-				sizeof(hdr->volume_serial),
-				"%016d", le32_to_cpu(hdr->volume_serial));
-	}
-
 	if (buf_fve) {
 		const struct bde_fve_metadata *fve = (const struct bde_fve_metadata *) buf_fve;
 
-		blkid_probe_sprintf_version(pr, "%d", fve->version);
+		blkid_probe_sprintf_version(pr, "%d", le16_to_cpu(fve->block_header.version));
```
Comment (t-8ch) [inline libblkid/src/superblocks/bitlocker.c:196]: Before your change, this was broken on big-endian, correct?
If so, could you add these fixes as extra commits before the introduction of the testcase.
This way they can be backported.

Comment (viccie30) [reply libblkid/src/superblocks/bitlocker.c:196]: Will do!

**Review by t-8ch — APPROVED** [review]

**Comment by karelzak** [comment]
@viccie30, it seems Coverity scan does not like the recent changes. I think it's false positives, but just for the record:
```
New defect(s) Reported-by: Coverity Scan
Showing 2 of 2 defect(s)


** CID 440626:  Insecure data handling  (TAINTED_SCALAR)
/libblkid/src/superblocks/bitlocker.c: 198 in probe_bitlocker()


________________________________________________________________________________________________________
*** CID 440626:  Insecure data handling  (TAINTED_SCALAR)
/libblkid/src/superblocks/bitlocker.c: 198 in probe_bitlocker()
192
193             if (buf_fve) {
194                     const struct bde_fve_metadata *fve = (const struct bde_fve_metadata *) buf_fve;
195
196                     blkid_probe_sprintf_version(pr, "%d", le16_to_cpu(fve->block_header.version));
197
>>>     CID 440626:  Insecure data handling  (TAINTED_SCALAR)
>>>     Using tainted variable "(uint32_t)__uint32_identity(fve->header.size)" as a loop boundary.
198                     for (off = sizeof(struct bde_fve_metadata_header);
199                          off + sizeof(struct bde_fve_metadata_entry) < le32_to_cpu(fve->header.size);
200                          off += le16_to_cpu(entry->size)) {
201                             entry = (const struct bde_fve_metadata_entry *) ((const char *) &fve->header + off);
202                             if (off % 2 ||
203                                 le16_to_cpu(entry->size) < sizeof(struct bde_fve_metadata_entry) ||

** CID 440625:  Memory - illegal accesses  (OVERRUN)
/libblkid/src/superblocks/bitlocker.c: 202 in probe_bitlocker()


________________________________________________________________________________________________________
*** CID 440625:  Memory - illegal accesses  (OVERRUN)
/libblkid/src/superblocks/bitlocker.c: 202 in probe_bitlocker()
196                     blkid_probe_sprintf_version(pr, "%d", le16_to_cpu(fve->block_header.version));
197
198                     for (off = sizeof(struct bde_fve_metadata_header);
199                          off + sizeof(struct bde_fve_metadata_entry) < le32_to_cpu(fve->header.size);
200                          off += le16_to_cpu(entry->size)) {
201                             entry = (const struct bde_fve_metadata_entry *) ((const char *) &fve->header + off);
>>>     CID 440625:  Memory - illegal accesses  (OVERRUN)
>>>     Overrunning array of 48 bytes at byte offset 48 by dereferencing pointer "entry".
202                             if (off % 2 ||
203                                 le16_to_cpu(entry->size) < sizeof(struct bde_fve_metadata_entry) ||
204                                 off + le16_to_cpu(entry->size) > le32_to_cpu(fve->header.size))
205                                     return -1;
206
207                             if (le16_to_cpu(entry->entry_type) == BDE_METADATA_ENTRY_TYPE_DESCRIPTION &&


```

**Pull request merged**